File: first_page_module/src/main/java/com/phone/first_page_module/FirstPageFragment.java
package com.phone.first_page_module;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.alibaba.android.arouter.launcher.ARouter;
import com.amap.api.location.AMapLocation;
import com.phone.common_library.BaseApplication;
import com.phone.common_library.base.BaseMvpRxFragment;
import com.phone.common_library.base.IBaseView;
import com.phone.common_library.bean.FirstPageResponse;
import com.phone.common_library.callback.OnCommonRxPermissionsCallback;
import com.phone.common_library.callback.OnCommonSingleParamCallback;
import com.phone.common_library.callback.OnItemViewClickListener;
import com.phone.common_library.manager.LogManager;
import com.phone.common_library.manager.RetrofitManager;
import com.phone.common_library.manager.RxPermissionsManager;
import com.phone.common_library.manager.ScreenManager;
import com.phone.common_library.manager.SystemManager;
import com.phone.common_library.service.IFirstPageService;
import com.phone.common_library.service.ISquareService;
import com.phone.common_library.ui.WebViewActivity;
import com.phone.first_page_module.adapter.FirstPageAdapter;
import com.phone.first_page_module.manager.AMAPLocationManager;
import com.phone.first_page_module.presenter.FirstPagePresenterImpl;
import com.phone.first_page_module.view.IFirstPageView;
import com.qmuiteam.qmui.widget.QMUILoadingView;
import com.scwang.smart.refresh.layout.SmartRefreshLayout;
import com.scwang.smart.refresh.layout.api.RefreshLayout;
import com.scwang.smart.refresh.layout.listener.OnRefreshLoadMoreListener;

import java.util.List;

/**
 * author    : Urasaki
 * e-mail    : 1164688204@qq.com
 * date      :
 * introduce :
 */
@Route(path = "/first_page_module/first_page")
public class FirstPageFragment extends BaseMvpRxFragment<IBaseView, FirstPagePresenterImpl>
        implements IFirstPageView {

    private static final String TAG = FirstPageFragment.class.getSimpleName();
    private FrameLayout layoutOutLayer;
    private Toolbar toolbar;
    private TextView tevTitle;
    private TextView tevRequestPermissionAndStartLocating;
    private SmartRefreshLayout refreshLayout;
    private RecyclerView rcvData;
    private QMUILoadingView loadView;

    private FirstPageAdapter firstPageAdapter;
    private boolean isRefresh;

    private AlertDialog mPermissionsDialog;
    private AMAPLocationManager amapLocationManager;

    private final String[] permissions = new String[]{
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.ACCESS_FINE_LOCATION};

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        // TODO: inflate a fragment view
        rootView = super.onCreateView(inflater, container, savedInstanceState);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    protected int initLayoutId() {
        return R.layout.fragment_first_page;
    }

    @Override
    protected void initData() {
        isRefresh = true;

        amapLocationManager = AMAPLocationManager.getInstance(BaseApplication.getInstance());
        amapLocationManager.setOnCommonSingleParamCallback(new OnCommonSingleParamCallback<AMapLocation>() {
            @Override
            public void onSuccess(AMapLocation success) {
                LogManager.i(TAG, "address*****" + success.getAddress());
            }

            @Override
            public void onError(String error) {
                LogManager.i(TAG, "error*****" + error);
            }
        });

    }

    @Override
    protected void initViews() {
        layoutOutLayer = (FrameLayout) rootView.findViewById(R.id.layout_out_layer);
        toolbar = (Toolbar) rootView.findViewById(R.id.toolbar);
        tevTitle = (TextView) rootView.findViewById(R.id.tev_title);
        tevRequestPermissionAndStartLocating = (TextView) rootView.findViewById(R.id.tev_request_permission_and_start_locating);
        refreshLayout = (SmartRefreshLayout) rootView.findViewById(R.id.refresh_layout);
        rcvData = (RecyclerView) rootView.findViewById(R.id.rcv_data);
        loadView = (QMUILoadingView) rootView.findViewById(R.id.loadView);

        tevTitle.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LogManager.i(TAG, "tev_title");
                initFirstPage();
            }
        });
        tevRequestPermissionAndStartLocating.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LogManager.i(TAG, "tevRequestPermissions");
                initRxPermissionsRxFragment();
            }
        });

        initAdapter();
    }

    private void initAdapter() {
        //	private FirstPageAdapter2                              firstPageAdapter;
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(rxAppCompatActivity);
        linearLayoutManager.setOrientation(RecyclerView.VERTICAL);
        rcvData.setLayoutManager(linearLayoutManager);
        rcvData.setItemAnimator(new DefaultItemAnimator());

        firstPageAdapter = new FirstPageAdapter(rxAppCompatActivity);
        //		firstPageAdapter = new FirstPageAdapter2(rxAppCompatActivity, R.layout.item_first_page);
        firstPageAdapter.setRcvOnItemViewClickListener(new OnItemViewClickListener() {
            @Override
            public void onItemClickListener(int position, View view) {
                //				if (view.getId() == R.id.tev_data) {
                //					//					url = "http://rbv01.ku6.com/omtSn0z_PTREtneb3GRtGg.mp4";
                //					//					url = "http://rbv01.ku6.com/7lut5JlEO-v6a8K3X9xBNg.mp4";
                //					url = "https://t-cmcccos.cxzx10086.cn/statics/shopping/detective_conan_japanese.mp4";
                //					//					fileFullname = mFileDTO.getFName();
                //					String[] arr = url.split("\\.");
                //					if (arr != null && arr.length > 0) {
                //						String fileName = "";
                //						StringBuilder stringBuilder = new StringBuilder();
                //						for (int i = 0; i < arr.length - 1; i++) {
                //							stringBuilder.append(arr[i]);
                //						}
                //						fileName = stringBuilder.toString();
                //						String suffix = arr[arr.length - 1];
                //
                //						paramMap.clear();
                //						paramMap.put("url", url);
                //						paramMap.put("suffix", suffix);
                //						startActivityCarryParams(ShowVideoActivity.class, paramMap);
                //					}
                //
                //				}

                if (view.getId() == R.id.ll_root) {
                    Intent intent = new Intent(rxAppCompatActivity, WebViewActivity.class);
                    intent.putExtra("loadUrl", firstPageAdapter.mJuheNewsBeanList.get(position).getUrl());
                    startActivity(intent);
                }
            }
        });
        rcvData.setAdapter(firstPageAdapter);

        refreshLayout.setOnRefreshLoadMoreListener(new OnRefreshLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                LogManager.i(TAG, "onLoadMore");
                isRefresh = false;
                initFirstPage();
            }

            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                isRefresh = true;
                initFirstPage();
            }
        });
    }

    @Override
    protected void initLoadData() {
        refreshLayout.autoRefresh();

        ISquareService ISquareService =
                (ISquareService) ARouter.getInstance().build("/square_module/SquareServiceImpl")
                        .navigation();
        LogManager.i(TAG,
                "squareService.getSquareDataList()******" + ISquareService.getSquareDataList().toString());

        LogManager.i(TAG, "FirstPageFragment initLoadData");

        //		startAsyncTask();

//        try {
//            //製造一個不會造成App崩潰的異常（类强制转换异常java.lang.ClassCastException）
//            User user = new User2();
//            User3 user3 = (User3) user;
//            LogManager.i(TAG, user3.toString());
//        } catch (Exception e) {
//            ExceptionManager.getInstance().throwException(rxAppCompatActivity, e);
//        }
    }

    @Override
    protected FirstPagePresenterImpl attachPresenter() {
        return new FirstPagePresenterImpl(this);
    }

//	private void startAsyncTask() {
//
//		// This async task is an anonymous class and therefore has a hidden reference to the outer
//		// class MainActivity. If the activity gets destroyed before the task finishes (e.g. rotation),
//		// the activity instance will leak.
//		new AsyncTask<Void, Void, Void>() {
//			@Override
//			protected Void doInBackground(Void... params) {
//				// Do some slow work in background
//				SystemClock.sleep(10000);
//				return null;
//			}
//		}.execute();
//
//		Toast.makeText(rxAppCompatActivity, "请关闭这个A完成泄露", Toast.LENGTH_SHORT).show();
//	}

    @Override
    public void showLoading() {
        if (loadView != null && !loadView.isShown()) {
            loadView.setVisibility(View.VISIBLE);
            loadView.start();
        }
    }

    @Override
    public void hideLoading() {
        if (loadView != null && loadView.isShown()) {
            loadView.stop();
            loadView.setVisibility(View.GONE);
        }
    }

    @Override
    public void firstPageDataSuccess(List<FirstPageResponse.ResultData.JuheNewsBean> success) {
        if (!rxAppCompatActivity.isFinishing()) {
            IFirstPageService IFirstPageService = (IFirstPageService) ARouter.getInstance().build("/first_page_module/FirstPageServiceImpl").navigation();
            if (isRefresh) {
                firstPageAdapter.clearData();
                firstPageAdapter.addData(success);
                refreshLayout.finishRefresh();
            } else {
                firstPageAdapter.addData(success);
                refreshLayout.finishLoadMore();
            }
            IFirstPageService.setFirstPageDataList(firstPageAdapter.mJuheNewsBeanList);
            hideLoading();
        }
    }

    @Override
    public void firstPageDataError(String error) {
        if (!rxAppCompatActivity.isFinishing()) {
            //            showToast(error, true);
            showCustomToast(ScreenManager.dpToPx(rxAppCompatActivity, 20f), ScreenManager.dpToPx(rxAppCompatActivity, 20f),
                    18, getResources().getColor(R.color.white),
                    getResources().getColor(R.color.color_FFE066FF), ScreenManager.dpToPx(rxAppCompatActivity, 40f),
                    ScreenManager.dpToPx(rxAppCompatActivity, 20f), error,
                    true);
            if (isRefresh) {
                refreshLayout.finishRefresh(false);
            } else {
                refreshLayout.finishLoadMore(false);
            }
            hideLoading();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 207) {
//            initRxPermissions();
        }
    }

    /**
     * RxFragment里需要的时候直接调用就行了
     */
    private void initRxPermissionsRxFragment() {
        RxPermissionsManager rxPermissionsManager = RxPermissionsManager.getInstance();
        rxPermissionsManager.initRxPermissions2(this, permissions, new OnCommonRxPermissionsCallback() {
            @Override
            public void onRxPermissionsAllPass() {
                //所有的权限都授予
//                //製造一個不會造成App崩潰的異常（类强制转换异常java.lang.ClassCastException）
//                User user = new User2();
//                User3 user3 = (User3) user;
//                LogManager.i(TAG, user3.toString());

                if (TextUtils.isEmpty(baseApplication.getSystemId())) {
                    String systemId = SystemManager.getSystemId(baseApplication);
                    baseApplication.setSystemId(systemId);
                    LogManager.i(TAG, "isEmpty systemId*****" + baseApplication.getSystemId());
                } else {
                    LogManager.i(TAG, "systemId*****" + baseApplication.getSystemId());
                }

                amapLocationManager.startLocation();
            }

            @Override
            public void onNotCheckNoMorePromptError() {
                //至少一个权限未授予且未勾选不再提示
                showSystemSetupDialog();
            }

            @Override
            public void onCheckNoMorePromptError() {
                //至少一个权限未授予且勾选了不再提示
                showSystemSetupDialog();
            }
        });
    }

    private void showSystemSetupDialog() {
        cancelPermissionsDialog();
        if (mPermissionsDialog == null) {
            mPermissionsDialog = new AlertDialog.Builder(rxAppCompatActivity)
                    .setTitle("权限设置")
                    .setMessage("获取相关权限失败，将导致部分功能无法正常使用，请到设置页面手动授权")
                    .setPositiveButton("去授权", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            cancelPermissionsDialog();
                            intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                            Uri uri = Uri.fromParts("package", rxAppCompatActivity.getApplicationContext().getPackageName(), null);
                            intent.setData(uri);
                            startActivityForResult(intent, 207);
                        }
                    })
                    .create();
        }

        mPermissionsDialog.setCancelable(false);
        mPermissionsDialog.setCanceledOnTouchOutside(false);
        mPermissionsDialog.show();
    }

    /**
     * 关闭对话框
     */
    private void cancelPermissionsDialog() {
        if (mPermissionsDialog != null) {
            mPermissionsDialog.cancel();
            mPermissionsDialog = null;
        }
    }

    private void initFirstPage() {
        showLoading();
        if (RetrofitManager.isNetworkAvailable(rxAppCompatActivity)) {
            bodyParams.clear();

            bodyParams.put("type", "yule");
            bodyParams.put("key", "d5cc661633a28f3cf4b1eccff3ee7bae");
            presenter.firstPage(this, bodyParams);
        } else {
            firstPageDataError(getResources().getString(R.string.please_check_the_network_connection));
        }
    }

    @Override
    public void onDestroyView() {
        if (layoutOutLayer != null) {
            layoutOutLayer.removeAllViews();
            layoutOutLayer = null;
        }
        if (amapLocationManager != null) {
            amapLocationManager.destoryLocation();
        }
        super.onDestroyView();
    }

}

File: module_home/src/main/java/com/phone/module_home/FirstPageFragment.kt
package com.phone.module_home

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.text.TextUtils
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.android.arouter.facade.annotation.Route
import com.alibaba.android.arouter.launcher.ARouter
import com.amap.api.location.AMapLocation
import com.phone.library_common.base.BaseMvpRxFragment
import com.phone.library_common.base.IBaseView
import com.phone.library_common.bean.FirstPageResponse.ResultData.JuheNewsBean
import com.phone.library_common.callback.OnCommonRxPermissionsCallback
import com.phone.library_common.callback.OnCommonSingleParamCallback
import com.phone.library_common.manager.*
import com.phone.library_common.service.IFirstPageService
import com.phone.library_common.service.ISquareService
import com.phone.library_common.ui.WebViewActivity
import com.phone.module_home.adapter.FirstPageAdapter
import com.phone.module_home.manager.AMAPLocationManager
import com.phone.module_home.presenter.FirstPagePresenterImpl
import com.phone.module_home.view.IFirstPageView
import com.qmuiteam.qmui.widget.QMUILoadingView
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnRefreshLoadMoreListener

@Route(path = "/module_home/home")
class FirstPageFragment : BaseMvpRxFragment<IBaseView, FirstPagePresenterImpl>(), IFirstPageView {

    private val TAG = FirstPageFragment::class.java.simpleName
    private var layoutOutLayer: FrameLayout? = null
    private var toolbar: Toolbar? = null
    private var tevTitle: TextView? = null
    private var tevRequestPermissionAndStartLocating: TextView? = null
    private var refreshLayout: SmartRefreshLayout? = null
    private var rcvData: RecyclerView? = null
    private var loadView: QMUILoadingView? = null

    private var firstPageAdapter: FirstPageAdapter? = null
    private var isRefresh = false

    private var mPermissionsDialog: AlertDialog? = null
    private var amapLocationManager: AMAPLocationManager? = null

    private val permissions = arrayOf(
        Manifest.permission.READ_EXTERNAL_STORAGE,
        Manifest.permission.WRITE_EXTERNAL_STORAGE,
        Manifest.permission.READ_PHONE_STATE,
        Manifest.permission.ACCESS_COARSE_LOCATION,
        Manifest.permission.ACCESS_FINE_LOCATION
    )

    override fun initLayoutId() = R.layout.fragment_first_page

    override fun initData() {
        isRefresh = true
        amapLocationManager = AMAPLocationManager.get()
        amapLocationManager?.setOnCommonSingleParamCallback(object :
            OnCommonSingleParamCallback<AMapLocation> {
            override fun onSuccess(success: AMapLocation) {
                SharedPreferencesManager.put("address", success.address)
                LogManager.i(TAG, "address*****" + SharedPreferencesManager.get("address", ""))
            }

            override fun onError(error: String) {
                LogManager.i(TAG, "error*****$error")
            }
        })
    }

    override fun initViews() {
        layoutOutLayer = rootView?.findViewById<View>(R.id.layout_out_layer) as FrameLayout
        toolbar = rootView?.findViewById<View>(R.id.toolbar) as Toolbar
        tevTitle = rootView?.findViewById<View>(R.id.tev_title) as TextView
        tevRequestPermissionAndStartLocating =
            rootView?.findViewById<View>(R.id.tev_request_permission_and_start_locating) as TextView
        refreshLayout = rootView?.findViewById<View>(R.id.refresh_layout) as SmartRefreshLayout
        rcvData = rootView?.findViewById<View>(R.id.rcv_data) as RecyclerView
        loadView = rootView?.findViewById<View>(R.id.loadView) as QMUILoadingView
        tevRequestPermissionAndStartLocating?.setOnClickListener {
            val ISquareService = ARouter.getInstance().build("/module_square/SquareServiceImpl")
                .navigation() as ISquareService
            LogManager.i(
                TAG,
                "squareService.getSquareDataList()******" + ISquareService.mSquareDataList.toString()
            )

            LogManager.i(TAG, "tevRequestPermissions")
            initRxPermissionsRxFragment()
        }
        initAdapter()
    }

    private fun initAdapter() {
        val linearLayoutManager = LinearLayoutManager(mRxAppCompatActivity)
        linearLayoutManager.orientation = RecyclerView.VERTICAL
        rcvData?.layoutManager = linearLayoutManager
        rcvData?.itemAnimator = DefaultItemAnimator()
        firstPageAdapter = FirstPageAdapter(mRxAppCompatActivity)
        firstPageAdapter?.setOnItemViewClickListener { position, view -> //				if (view.getId() == R.id.tev_data) {
            //					//					url = "http://rbv01.ku6.com/omtSn0z_PTREtneb3GRtGg.mp4"
            //					//					url = "http://rbv01.ku6.com/7lut5JlEO-v6a8K3X9xBNg.mp4"
            //					url = "https://t-cmcccos.cxzx10086.cn/statics/shopping/detective_conan_japanese.mp4"
            //					//					fileFullname = mFileDTO.getFName()
            //					String[] arr = url.split("\\.")
            //					if (arr != null && arr.length > 0) {
            //						String fileName = ""
            //						StringBuilder stringBuilder = new StringBuilder()
            //						for (int i = 0 i < arr.length - 1 i++) {
            //							stringBuilder.append(arr[i])
            //						}
            //						fileName = stringBuilder.toString()
            //						String suffix = arr[arr.length - 1]
            //
            //						paramMap.clear()
            //						paramMap.put("url", url)
            //						paramMap.put("suffix", suffix)
            //						startActivityCarryParams(ShowVideoActivity.class, paramMap)
            //					}
            //
            //				}
            if (view.id == R.id.ll_root) {
                val intent = Intent(mRxAppCompatActivity, WebViewActivity::class.java)
                intent.putExtra(
                    "loadUrl",
                    firstPageAdapter?.mJuheNewsBeanList?.get(position)?.url
                )
                startActivity(intent)
            }
        }
        rcvData?.adapter = firstPageAdapter
        refreshLayout?.setOnRefreshLoadMoreListener(object : OnRefreshLoadMoreListener {
            override fun onLoadMore(refreshLayout: RefreshLayout) {
                LogManager.i(TAG, "onLoadMore")
                isRefresh = false
                initFirstPage()
            }

            override fun onRefresh(refreshLayout: RefreshLayout) {
                isRefresh = true
                initFirstPage()
            }
        })
    }

    override fun initLoadData() {
        initFirstPage()
        LogManager.i(TAG, "FirstPageFragment initLoadData")

        //		startAsyncTask()

//        try {
//            //製造一個不會造成App崩潰的異常（类强制转换异常java.lang.ClassCastException）
//            User user = new User2()
//            User3 user3 = (User3) user
//            LogManager.i(TAG, user3.toString())
//        } catch (Exception e) {
//            ExceptionManager.getInstance().throwException(getRxAppCompatActivity(), e)
//        }
    }

    override fun attachPresenter(): FirstPagePresenterImpl {
        return FirstPagePresenterImpl(this)
    }

//	private void startAsyncTask() {
//
//		// This async task is an anonymous class and therefore has a hidden reference to the outer
//		// class MainActivity. If the activity gets destroyed before the task finishes (e.g. rotation),
//		// the activity instance will leak.
//		new AsyncTask<Void, Void, Void>() {
//			@Override
//			protected Void doInBackground(Void... params) {
//				// Do some slow work in background
//				SystemClock.sleep(10000)
//				return null
//			}
//		}.execute()
//
//		Toast.makeText(getRxAppCompatActivity(), "请关闭这个A完成泄露", Toast.LENGTH_SHORT).show()
//	}

    //	private void startAsyncTask() {
    //
    //		// This async task is an anonymous class and therefore has a hidden reference to the outer
    //		// class MainActivity. If the activity gets destroyed before the task finishes (e.g. rotation),
    //		// the activity instance will leak.
    //		new AsyncTask<Void, Void, Void>() {
    //			@Override
    //			protected Void doInBackground(Void... params) {
    //				// Do some slow work in background
    //				SystemClock.sleep(10000)
    //				return null
    //			}
    //		}.execute()
    //
    //		Toast.makeText(getRxAppCompatActivity(), "请关闭这个A完成泄露", Toast.LENGTH_SHORT).show()
    //	}

    override fun showLoading() {
        loadView?.let {
            if (it.isShown) {
                it.visibility = View.VISIBLE
                it.start()
            }
        }
    }

    override fun hideLoading() {
        loadView?.let {
            if (it.isShown) {
                it.stop()
                it.visibility = View.GONE
            }
        }
    }

    override fun firstPageDataSuccess(success: List<JuheNewsBean>) {
        if (!mRxAppCompatActivity.isFinishing) {
            if (isRefresh) {
                firstPageAdapter?.clearData()
                firstPageAdapter?.addData(success)
                refreshLayout?.finishRefresh()
            } else {
                firstPageAdapter?.addData(success)
                refreshLayout?.finishLoadMore()
            }
            LogManager.i(
                TAG,
                "firstPageAdapter?.mJuheNewsBeanList*****" + firstPageAdapter?.mJuheNewsBeanList.toString()
            )

            val IFirstPageService =
                ARouter.getInstance().build("/module_home/FirstPageServiceImpl")
                    .navigation() as IFirstPageService
            IFirstPageService.mFirstPageDataList =
                firstPageAdapter?.mJuheNewsBeanList ?: mutableListOf()
            hideLoading()
        }
    }

    override fun firstPageDataError(error: String) {
        if (!mRxAppCompatActivity.isFinishing) {
//            showToast(error, true)
            showCustomToast(
                ScreenManager.dpToPx(20f), ScreenManager.dpToPx(20f),
                18, resources.getColor(R.color.white),
                resources.getColor(R.color.color_FF198CFF), ScreenManager.dpToPx(40f),
                ScreenManager.dpToPx(20f), error,
                true
            )
            if (isRefresh) {
                refreshLayout?.finishRefresh(false)
            } else {
                refreshLayout?.finishLoadMore(false)
            }
            hideLoading()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 207) {
//            initRxPermissions()
        }
    }

    /**
     * RxFragment里需要的时候直接调用就行了
     */
    private fun initRxPermissionsRxFragment() {
        val rxPermissionsManager = RxPermissionsManager.instance()
        rxPermissionsManager.initRxPermissions2(
            this,
            permissions,
            object : OnCommonRxPermissionsCallback {
                override fun onRxPermissionsAllPass() {
                    //所有的权限都授予
//                //製造一個不會造成App崩潰的異常（类强制转换异常java.lang.ClassCastException）
//                User user = new User2()
//                User3 user3 = (User3) user
//                LogManager.i(TAG, user3.toString())
                    val systemId = SharedPreferencesManager.get("systemId", "") as String
                    if (TextUtils.isEmpty(systemId)) {
                        SharedPreferencesManager.put("systemId", SystemManager.getSystemId())
                        LogManager.i(TAG,
                            "isEmpty systemId*****${SharedPreferencesManager.get("systemId", "") as String}")
                    } else {
                        LogManager.i(TAG, "systemId*****$systemId")
                    }
                    amapLocationManager?.startLocation()
                }

                override fun onNotCheckNoMorePromptError() {
                    //至少一个权限未授予且未勾选不再提示
                    showSystemSetupDialog()
                }

                override fun onCheckNoMorePromptError() {
                    //至少一个权限未授予且勾选了不再提示
                    showSystemSetupDialog()
                }
            })
    }

    private fun showSystemSetupDialog() {
        cancelPermissionsDialog()
        if (mPermissionsDialog == null) {
            mPermissionsDialog = AlertDialog.Builder(mRxAppCompatActivity)
                .setTitle("权限设置")
                .setMessage("获取相关权限失败，将导致部分功能无法正常使用，请到设置页面手动授权")
                .setPositiveButton("去授权") { dialog, which ->
                    cancelPermissionsDialog()
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                    val uri = Uri.fromParts(
                        "package", mBaseApplication.packageName, null
                    )
                    intent.data = uri
                    startActivityForResult(intent, 207)
                }
                .create()
        }
        mPermissionsDialog?.setCancelable(false)
        mPermissionsDialog?.setCanceledOnTouchOutside(false)
        mPermissionsDialog?.show()
    }

    /**
     * 关闭对话框
     */
    private fun cancelPermissionsDialog() {
        mPermissionsDialog?.cancel()
        mPermissionsDialog = null
    }

    private fun initFirstPage() {
        showLoading()
        if (RetrofitManager.isNetworkAvailable()) {
            mBodyParams.clear()
            mBodyParams["type"] = "yule"
            mBodyParams["key"] = "d5cc661633a28f3cf4b1eccff3ee7bae"
            presenter?.firstPage(this, mBodyParams)
        } else {
            firstPageDataError(resources.getString(R.string.please_check_the_network_connection))
        }
    }

    override fun onDestroyView() {
        layoutOutLayer?.removeAllViews()
        layoutOutLayer = null
        if (amapLocationManager != null) {
            amapLocationManager?.destoryLocation()
        }
        super.onDestroyView()
    }

}